### ISSUE TITLE
Async methods excluded from coverage after updating to 2.8.1

### ISSUE THREAD

**Issue by Malivil** [tonerdo/coverlet#794]
Using coverage format "OpenCover" and the coverlet.msbuild package
In version 2.8.0 an async method in my Web API class shows no coverage (because I haven't written tests yet):
![BeforeCoverage](https://user-images.githubusercontent.com/1238511/78559264-16f9b800-77e2-11ea-9809-3064f4fab7f2.png)

In version 2.8.1, the same method is just excluded from coverage entirely:
![AfterCoverage](https://user-images.githubusercontent.com/1238511/78559283-211bb680-77e2-11ea-8bbe-8f4fab0933f0.png)

The coverage.opencover.xml shows a complete missing section for the method:
![Diff](https://user-images.githubusercontent.com/1238511/78559322-30026900-77e2-11ea-9b64-4c248c4b2d69.png)

I do not have a minimal reproduction example but if needed I can probably make one.
May be related to 528956bc1189eb541a6f7183819261f4b20e5dcc

**Comment by MarcoRossignoli** [comment]
Do you have `ExcludeFromCodeCoverage` attribute somewhere?
Do you see missing coverage only for this method in source or for all method of same class?

cc: @matteoerigozzi

**Comment by Malivil** [comment]
This class does not have `ExcludeFromCodeCoverage`
Only the two methods marked `async` are not showing coverage.

I will try to make a simple reproduction project. I'll comment here with the repo when it's complete.

**Comment by MarcoRossignoli** [comment]
Ok thanks a lot!

**Comment by Malivil** [comment]
Here's a repro with the minimum example: https://github.com/Malivil/CoverletAsyncRepro
If you run the .bat file within the .Tests project it will generate a report in the TestResults folder

Changing the version in the .Tests csproj to 2.8.0 and then re-running the .bat will show the function as uncovered as it should.

**Comment by MarcoRossignoli** [comment]
Great, stay on `2.8.0` for now I'll come back asap.

**Comment by bengoymer** [comment]
Hi there,

I've also noticed this issue recently, it appears to be something to do with the `/p:ExcludeByAttribute=CompilerGeneratedAttribute` part, as removing this will cause the coverage to be successfully reported. Thanks for the tip of going back to 2.8.0 as that also solves the issue for now.

**Comment by MarcoRossignoli** [comment]
@Malivil can you confirm that removing `/p:ExcludeByAttribute=CompilerGeneratedAttribute` solve the issue?

**Comment by bengoymer** [comment]
@MarcoRossignoli I feel that I need to clarify; athough removing the argument appears to solve this issue, it will cause all our auto-implemented properties in our DTOs to record no coverage without being explictly tested, which we did not have to do prior to 2.8.1.

**Comment by MarcoRossignoli** [comment]
@bengoymer yes I mean this is a bug but if we know that is related to that parameter we can fix more fast.

**Comment by bengoymer** [comment]
@MarcoRossignoli perfect, thank you.

**Comment by MarcoRossignoli** [comment]
@bengoymer can you try to remove that param and filter out DTO using source file filtering(if DTO are in separate files)?
https://github.com/tonerdo/coverlet/blob/master/Documentation/MSBuildIntegration.md#source-files

**Comment by bengoymer** [comment]
@MarcoRossignoli yes, just done that. If I filter out the DTOs using the source file filtering, and remove ExcludeByAttribute the coverage of the async functions is as I'd expect.

**Comment by Malivil** [comment]
@MarcoRossignoli Removing that exclude does fix this issue. The reason I had that was mostly to exclude generated property Getters and Setters.

**Comment by MarcoRossignoli** [comment]
@Malivil sorry I wasn't clear, this is only to confirm that the issue disappear if you remove that attribute for those async method, it won't fix the issue it's to have a double confirmation on where issue could be.

**Comment by Malivil** [comment]
@MarcoRossignoli I understand. I was just adding clarity as to why I had it there in the first place =)

**Comment by MarcoRossignoli** [comment]
So if you remove that attribute you see async methods uncovered yet?

**Comment by Malivil** [comment]
Yes, removing the attribute exclude shows the async methods as being uncovered, that's what I meant when I said it "[fixed] the issue". Sorry for not being clear =)

**Comment by viceice** [comment]
Seeing this too, workaround:

```xml
<!-- <ExcludeByAttribute>Obsolete,GeneratedCodeAttribute,CompilerGeneratedAttribute</ExcludeByAttribute> -->
<ExcludeByAttribute>Obsolete</ExcludeByAttribute>
```

**Comment by MarcoRossignoli** [comment]
@Malivil can you try the workaround(temporary) and tell me if it works?

**Comment by Malivil** [comment]
Where does that go? In the test .csproj?

**Comment by viceice** [comment]
@Malivil It depends on your usage. We use the `collector`, so we add this to the `runsettings` file.

**Comment by Malivil** [comment]
We use the msbuild integration as shown in the test repo: https://github.com/Malivil/CoverletAsyncRepro

**Comment by viceice** [comment]
You need to add it to your test csproj.

https://github.com/tonerdo/coverlet/blob/master/Documentation/MSBuildIntegration.md

**Comment by Malivil** [comment]
That's what I thought.

@MarcoRossignoli: This workaround behaves the same as the command-line change which I think is expected.

**Comment by MarcoRossignoli** [comment]
> This workaround behaves the same as the command-line change which I think is expected.

Not clear what you mean,  `/p:ExcludeByAttribute=Obsolete` for msbuild

**Comment by Malivil** [comment]
> > This workaround behaves the same as the command-line change which I think is expected.
> 
> Not clear what you mean, `/p:ExcludeByAttribute=Obsolete` for msbuild

Using that switch behaves the same as if I don't use `ExcludeByAttribute` at all. I don't have anything annotated with `Obsolete` so it has no affect.

**Comment by MarcoRossignoli** [comment]
> Using that switch behaves the same as if I don't use ExcludeByAttribute at all. I don't have anything annotated with Obsolete so it has no affect.

Yes the test was only to understand if we're excluding something generated by compiler and also with `Obsolete` and maybe the result could be a bit different. But ok thanks for help on investigation.

**Comment by Malivil** [comment]
If I use `Obsolete`, `GeneratedCodeAttribute`, or a combination of both it does not exclude the async method. If I add `CompilerGeneratedAttribute` back in to either the switch or the test .csproj then it does exclude the async method again. So it seems to be narrowed to `CompilerGeneratedAttribute` specifically.

**Comment by MarcoRossignoli** [comment]
> So it seems to be narrowed to CompilerGeneratedAttribute specifically.

Yep, my hope was find a temporary workaround with no version downgrade, but seem not possible at the moment.

**Comment by MarcoRossignoli** [comment]
@Malivil  we did some check, actually the behaviour is correct, we fixed a bug where `CompilerGeneratedAttribute` wasn't honored, so now correct behaviour is to exclude class with that attribute and in your case every async state machine(autogenerated class by roslyn compiler for async methods) has got that attribute. 
You cannot use that filter to remove autoprops from instrumentation anymore and it's correct.
A solution for now is to use `ExcludeByFile` filter and remove al DTO source files.

We should add support for skip autoprops https://github.com/tonerdo/coverlet/issues/328
